packages/core/core/src/core/app-configuration/config-loader.ts

```diff
@@ -4,17 +4,112 @@ import { loadFile } from './load-config-file';
 
 const VALID_EXTENSIONS = ['.js', '.json'];
 
+// These filenames are restricted, but will also emit a warning that the filename is probably a mistake
+const MISTAKEN_FILENAMES = {
+  middleware: 'middlewares',
+  plugin: 'plugins',
+};
+
+// the following are restricted to prevent conflicts with existing STRAPI_* env vars or root level config options
+// must all be lowercase to match validator
+const RESTRICTED_FILENAMES = [
+  // existing env vars
+  'uuid',
+  'hosting',
+  'license',
+  'enforce',
+  'disable',
+  'enable',
+  'telemetry',
+
+  // reserved for future internal use
+  'strapi',
+  'internal',
+
+  // root level config options
+  // TODO: these might be better out of root config
+  'launchedAt',
+  'serveAdminPanel',
+  'autoReload',
+  'environment',
+  'packageJsonStrapi',
+  'info',
+  'autoReload',
+
+  // probably mistaken/typo filenames
+  ...Object.keys(MISTAKEN_FILENAMES),
+];
+
+const warnSkippedConfig = (message: string) => {
+  console.warn(message);
+};
+
 export default (dir: string) => {
   if (!fs.existsSync(dir)) return {};
 
-  return fs
-    .readdirSync(dir, { withFileTypes: true })
-    .filter((file) => file.isFile() && VALID_EXTENSIONS.includes(path.extname(file.name)))
-    .reduce((acc, file) => {
-      const key = path.basename(file.name, path.extname(file.name));
+  const allFiles = fs.readdirSync(dir, { withFileTypes: true });
+  const seenFilenames = new Set<string>();
+  const configFiles = allFiles.reduce((acc, file) => {
+    const baseName = path.basename(file.name, path.extname(file.name));
+    const baseNameLower = baseName.toLowerCase();
+    const extension = path.extname(file.name);
+    const extensionLower = extension.toLowerCase();
 
-      acc[key] = loadFile(path.resolve(dir, file.name));
+    if (!file.isFile()) {
+      return acc;
+    }
 
+    // Note: we don't have access to strapi logger at this point so we can't check if
+    if (!VALID_EXTENSIONS.includes(extensionLower)) {
+      warnSkippedConfig(
+        `Config file not loaded, extension must be one of ${VALID_EXTENSIONS.join(',')}): ${
+          file.name
+        }`
+      );
       return acc;
-    }, {} as Record<string, unknown>);
+    }
+
+    if (RESTRICTED_FILENAMES.includes(baseNameLower)) {
+      warnSkippedConfig(`Config file not loaded, restricted filename: ${file.name}`);
+
+      // suggest the filename they probably meant
+      if (baseNameLower in MISTAKEN_FILENAMES) {
+        console.log(
+          `Did you mean ${MISTAKEN_FILENAMES[baseNameLower as keyof typeof MISTAKEN_FILENAMES]}]} ?`
+        );
+      }
+
+      return acc;
+    }
+
+    // filter filenames with non-alphanumberic characters
+    // because underscore (_) is reserved as a separator for use when loading env vars into config
+    if (!/^[A-Za-z0-9]+$/.test(baseName)) {
```

Comment:
Does this mean that you can't have `content-types`? It seems a bit restrictive to forbid characters such as `-`, no?

Reply:
Yeah, I realized it might be a bit too restrictive. I think we can safely loosen it, we only need to be certain about our own files. I'll update it.

Reply:
Done